### ISSUE TITLE
PYJD and PYJS handle non-ascii characters differently in text widgets

### ISSUE THREAD

**Issue by pyjsorg** [pyjs/original-pyjs#42]
<b>What steps will reproduce the problem?</b>
1. Create a TextBox or TextArea widget
2. call widget.setText(somestring) where somestring contains a non-ascii character such as the UK pound character
3. Alternatively, type a string with a non-ascii character into the widget and then call widget.getText()

If you are running PYJS, it will work correctly. If you are running PYJD and exception will be thrown in both cases, giving the error message (for a UK pound character):

UnicodeEncodeError: 'ascii' codec can't encode character u'\xa3' in position 0: ordinal not in range(128)

The reasons are obvious:

The TextBoxBase implementation of setText(text) is
    DOM.setAttribute(self.getElement(), &quot;value&quot;, str(text))

for getText() it has:
    return DOM.getAttribute(self.getElement(), &quot;value&quot;)
but DOM.getAttribute(elem, attr) is:
    attr = getattr(elem, attr)
    if attr is None:
        return None
    return str(attr)

In both cases the text is being passed through the str() builtin function which in Python (hence in PYJD) will throw the observed exception if it encounters any non-ascii characters. 
But in Javascript, strings are UTF16 and apparently the implementation of the str() function in PYJS tolerates non-ascii characters.

Possible solution:

Change str() to unicode() in DOM.getAttribute and DOM.getElementAttribute() as appropriate and do the same in TextBoxBase.setText()

<b>What version of Pyjamas are you using?  which svn revision number?</b>
4353ada...
I know this is not the latest but it is very recent and I don't think any changes have been made in DOM or TextBox base in a long time.

<b>On what operating system?</b>
Windows 7

_Original issue: http://code.google.com/p/pyjamas/issues/detail?id=681 (February 02, 2012 16:39:09)_

**Comment by pyjsorg** [comment]
_From jpcharle...@gmail.com on February 05, 2012 13:01:36:_
I checked this problem on a Debian installation using hulahop and the problem did not occur, so it seems likely that it is confined to mshtml installations. In view of that, rather than modify ui/TextBoxBase.py and DOM.py, we have agreed that ui/TextBoxBase.mshtml.py and DOM.mshtml.py should be the target of any fixes.

The changes therefore comprise:

adding setText(text) to TextBoxBase.mshtml.py, identical with that in TextBoxBase.py   except for the fact that the text argument no longer has str() applied to it.

adding getAttribute(elem, attr) and getElemAttribute(elem, attr) to DOM.mshtml.py, identical to the versions in DOM.py except that all the str() calls have been removed. TextBoxBase,getText() only calls getAttribute, so the change to getElemAttribute is strictly not necessary but seems advisable in the circumstances.

A diff file for these changes is attached
